### ISSUE TITLE
Support python generation using `grpc_tools` python package

### ISSUE THREAD

**Issue by jlewi** [bufbuild/buf#1344]
# TL;DR

Would it possible to support generating python bindings by invoking the python package `grpc_tools` as opposed to relying on `protoc` and `protoc` plugins?

# Details

There have been a couple issues detailing how to use buf to generate Python protocol bindings. #223 and #1287. The root issue here appears to be that buf relies on `protoc-gen-python_grpc` being on the path. But the gRPC community doesn't support this way of generating python gRPC bindings.  The officially supported python generation path is using `grpc_tools` which compiles in the plugin as a python extension (see [comment](https://github.com/grpc/grpc/issues/15675#issuecomment-397038811) and [docs](https://grpc.io/docs/languages/python/basics/#generating-client-and-server-code).

As noted in #223 building `protoc-gen-python_grpc` is non-straightforward. Especially on Mac M1 (grpc/grpc#29262). I tried to build on my M1 and got `Unsupported architecture: Darwin arm64`. Wheras I was able to install `grpc_tools` and use that just fine.

If `buf` supports using `grpc_tools` to generate python bindings that would align it with what is officially supported by the python gRPC community. I expect this would eliminate a lot of headaches for folks. 

My naive assumption is that buf is just shelling out to protoc so hopefully shelling out to other tools is a viable path? Is that accurate?

# Alternative solutions

## remote packages

#223  suggests using remote plugins on the BSR service. Requiring private code to be shipped to an external compilation service isn't a great solution. 

## grpclib

grpclib ships some prebuilt versions of the plugins; I found these generate different code which use the grpclib then `grpc_tools`.

**Comment by amckinney** [comment]
Thanks for the issue! This one is slightly involved, so I'll address your questions independently.

> Would it possible to support generating python bindings by invoking the python package grpc_tools as opposed to relying on protoc and protoc plugins?

Short answer, no. I'll explain why by addressing each of your sub-points.

> The root issue here appears to be that buf relies on protoc-gen-python_grpc being on the path. But the gRPC community doesn't support this way of generating python gRPC bindings. The officially supported python generation path is using grpc_tools which compiles in the plugin as a python extension (see https://github.com/grpc/grpc/issues/15675#issuecomment-397038811 and [docs](https://grpc.io/docs/languages/python/basics/#generating-client-and-server-code).

More accurately, it's because the `protoc-gen-python-grpc` plugin _violates_ the traditional plugin interface - a stand-alone program that can be installed on your `$PATH`, and reads a `CodeGeneratorRequest` from stdin and writes a `CodeGeneratorResponse` to stdout. This is not a flaw of `buf` - it's a flaw of the `protoc-gen-python-grpc` plugin.

> My naive assumption is that buf is just shelling out to protoc so hopefully shelling out to other tools is a viable path? Is that accurate?

No that's not accurate - `buf` has its own [internal compiler](https://docs.buf.build/reference/internal-compiler) written in pure Go. Shelling out isn't an option here.

> https://github.com/bufbuild/buf/issues/223 suggests using remote plugins on the BSR service. Requiring private code to be shipped to an external compilation service isn't a great solution.

This is the best solution that we have today - our remote generation model makes it so that you don't have to worry about the manual installation / execution of the plugin locally (explained [here](https://github.com/bufbuild/buf/issues/223#issuecomment-834545444)).

### Ok, so now what?

With all that said, there are two functional solutions today: a less than ideal solution for invoking the plugin locally (as a result of the `grpc_tools` design hack), and another solution for generating the plugin remotely. I realize the remote generation approach might not suit your needs, so you'll need to rely on the local generation approach for now.

The right solution here would be to have `protoc-gen-python-grpc` refactored as a real `protoc` plugin (just like the rest).

**Comment by jlewi** [comment]
@amckinney Really appreciate the detailed explanation.

**Comment by emcfarlane** [comment]
For anyone coming to this old issue, have a look at the grpc/python plugin: https://buf.build/grpc/python?version=v1.62.1

**Comment by stevvooe** [comment]
Oof, super bizarre that grpc_tools doesn't implement the actual plugin protocol. While the remote plugin is handy for certain projects, I don't think we can send code to your service. Any way we can set the protoc binary to the grpc_tools.protoc call? It looks like buf is validating the existence of grpc_python. If it didn't do that, it could just farm the call out and we could make it work pretty easily.

Let me know and I'll even send a patch to fix this.
